### PR TITLE
feat: add click name service

### PR DESCRIPTION
**Pull request by aliXsed**
Useful info:
* `clk.eth` name info is:
```
Label hash: 0x3c5cc49c4c05caada7ffe8d159634bd3e1825c59241310705d45e578f33da648
Nodehash: 0x71691213bee783e965b983e04ea3aac38f8277868655e40363711821e5422940
```
* TokenId of `clk.eth` in [NameWrapper](https://etherscan.io/token/0xd4416b13d2b3a9abae7acd5d6c2bbdbe25686401?a=51296995595488636425956155272184315314353991018314322163154949517603913214272) is `51296995595488636425956155272184315314353991018314322163154949517603913214272` equal to `0x71691213bee783e965b983e04ea3aac38f8277868655e40363711821e5422940`
* [clk.eth](https://app.ens.domains/clk.eth) and `clickapp.eth` both are currently resolved to `0xaD56AaCEb54407803dDA93D6b1ba6b85676A2226`. This [name is registered and controlled](https://etherscan.io/name-lookup-search?id=clk.eth) by `0xD4416b13d2b3a9aBae7AcD5D6C2BbDBE25686401`
* clk.eth's resolver is the `PublicResolver` at [0x231b0Ee14048e9dCcD1d247744d114a4EB5E8E63](https://etherscan.io/address/0x231b0Ee14048e9dCcD1d247744d114a4EB5E8E63#code)
* Sepolia L1 DiamondProxy Contract for ZkSync-Era: `0x9A6DE0f62Aa270A8bCB1e2610078650D539B1Ef9`
* Sepolia L1 StorageVerifier Contract for ZkSync-Era: `0x5490D0FE20E9F93a847c1907f7Fd2adF217bF534`
* ETH L1 DiamondProxy Contract for ZkSync-Era: `0x32400084C286CF3E17e7B677ea9583e60a000324`
* Nodle's instance of SparseMerkleTree is deployed at [0xdc5503a345a584382ef9c8dcb015e5ec095c730c](https://sepolia.etherscan.io/address/0xdc5503a345a584382ef9c8dcb015e5ec095c730c#code)
* Nodle's instance of StorageProofVerifier is deployed at [0x8e8e263a394d5637a169424f1224c2da7ef406ad](https://sepolia.etherscan.io/address/0x8e8e263a394d5637a169424f1224c2da7ef406ad)
* Nodle's Test L2 ClickNameService deployed at [0xAD4360c87275C7De1229A8c3C0567f921c9302b1](https://sepolia.explorer.zksync.io/address/0xAD4360c87275C7De1229A8c3C0567f921c9302b1#contract)
* Here is the storage layout of ClickNameService:
```
┌───────────────────────┬─────────────────────────────────┬──────────────┬────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────┬─────┬───────────────────────────────────────────────────┬───────────────┐
│       contract        │         state_variable          │ storage_slot │ offset │                                                type                                                 │ idx │                     artifact                      │ numberOfBytes │
├───────────────────────┼─────────────────────────────────┼──────────────┼────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────┼─────┼───────────────────────────────────────────────────┼───────────────┤
│   ClickNameService    │              _name              │      0       │   0    │                                          t_string_storage                                           │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │             _symbol             │      1       │   0    │                                          t_string_storage                                           │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │             _owners             │      2       │   0    │                                   t_mapping(t_uint256,t_address)                                    │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │            _balances            │      3       │   0    │                                   t_mapping(t_address,t_uint256)                                    │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │         _tokenApprovals         │      4       │   0    │                                   t_mapping(t_uint256,t_address)                                    │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │       _operatorApprovals        │      5       │   0    │                          t_mapping(t_address,t_mapping(t_address,t_bool))                           │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │             _roles              │      6       │   0    │                         t_mapping(t_bytes32,t_struct(RoleData)302_storage)                          │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │         expiryDuration          │      7       │   0    │                                              t_uint256                                              │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
│   ClickNameService    │             expires             │      8       │   0    │                                   t_mapping(t_uint256,t_uint256)                                    │  0  │ /build-info/2dae3c66f2605f8c8b126e96e1bd78d1.json │      32       │
```
* Nodle's Test L1 ClickResolver is deployed at [0x49E1E847a75C00241596CCBdA8cbEE977be4e8eF](https://sepolia.etherscan.io/address/0x49E1E847a75C00241596CCBdA8cbEE977be4e8eF)
* An indexer for registered names is deployed at https://test-indexer.nodleprotocol.io/ which can reply to queries like below:
```
{
  account(id: "0x2e7f3926ae74fdcdcade2c2ab50990c5dafd42bd") {
    eNsByOwnerId {
      nodes {
        name
      }
    }
  }
}
```

TODO:
- [x] Find an instance of (or deploy) the StorageProofVerifier contract on Ethereum L1
- [x] If StorageProofVerifier needs to be deployed, find the right contract address for IZkSyncDiamond
- [x] Deploy an instance of ENSRegistery (for the base  name of `clk` or `clickapp`) on ZkSync, approve our backend as operator for subdomains.
- [x] Develop ClickResolver based on ClaveResolver and deploy it on L1
- [x] Set the ClickResolver address for the node of `aliXsed.eth` on L1 NameWrapper
- [x] Create the skeleton of a ts project to become the clk-gateway cloud run
- [x] Add L2 resolve API to clk-gateway
- [x] Add L2 register API to clk-gateway
- [x] Load configuration from .env
- [x] Dockerize and push docker to GC artifact registry
- [x] Deploy clk-gateway to our gc infra (https://github.com/NodleCode/infra/pull/991)
- [x] Validate all API inputs
- [x] Provide registerL1 API only to users with verified email 
- [x] Test the email verification flow for accessing POST /registerL1 
- [x] Add L1 resolve with full proof validation API to clk-gateway
- [x] Improve error handling to distinguish between gateway unknown error and user input error (predictable errors like name not registered when resolving)
- [x] Add L1 resolve (no proof validation) API to clk-gateway
- [x] Fix CI build
- [x] Consider removing toLower for domain check in ClickResolver or adding it to ClickNameService
- [x] Allow full CORS
- [x] Limit /registerL1 to only allow one name per user
- [x] Restrict names to be at least 5 letter longs
- [x] Address review comments
- [x] Add indexer query for reverse lookup an address and find all the names owned by that address
- [ ] Enable batch register to alleviate nonce issue for concurrent POST requests on /registerL2 (Deferred to subsequent PRs)
- [ ] Add paymaster option for registration ((Deferred to subsequent PRs)
- [ ] (Optional) Reserve some names (Deferred to be done as an admin job not in the code)

### CODE REVIEW

File: src/nameservice/ClickNameService.sol [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    constructor(address admin, address registrar) ERC721("ClickNameService", "CLK") {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(REGISTERER_ROLE, registrar);
+    }
+
+    /// @inheritdoc IClickNameService
+    function resolve(string memory name) external view returns (address) {
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = ownerOf(tokenId);
+        require(expires[tokenId] > block.timestamp, "Name expired.");
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:45]: We should revert with a named error like `error NameExpired()`

Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:45]: Expiration should be toggle-able as for now we do not want names to expire (yet)

File: src/nameservice/ClickNameService.sol [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        return owner;
+    }
+
+    /**
+     * @notice Register multiple names at once
+     */
+    function batchRegister(NameOwner[] memory nameOwners) external isAuthorized {
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:52]: Not a big fan of modifiers, we should go back to our normal pattern (this also applies to pretty much most functions below)

File: src/nameservice/ClickNameService.sol [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        emit NameRegistered(name, to, expireTimestamp);
+    }
+
+    /// @inheritdoc IClickNameService
+    function registerWithExpiry(address to, string memory name, uint256 duration) public isAuthorized {
+        uint256 tokenId = _register(to, name);
+        uint256 expireTimestamp = block.timestamp + duration;
+        expires[tokenId] = expireTimestamp;
+        emit NameRegistered(name, to, expireTimestamp);
+    }
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:80]: This should deduped like so:
```suggestion
    /// @inheritdoc IClickNameService
    function register(address to, string memory name) public isAuthorized {
       registerWithExpiry(to, name, expiryDuration);
    }

    /// @inheritdoc IClickNameService
    function registerWithExpiry(address to, string memory name, uint256 duration) public isAuthorized {
        uint256 tokenId = _register(to, name);
        uint256 expireTimestamp = block.timestamp + duration;
        expires[tokenId] = expireTimestamp;
        emit NameRegistered(name, to, expireTimestamp);
    }
```

File: src/nameservice/ClickNameService.sol [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    function _register(address to, string memory name) private returns (uint256) {
+        require(bytes(name).length != 0, "Name cannot be empty.");
+        require(_isAlphanumeric(name), "Name must be alphanumeric.");
+
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = _ownerOf(tokenId);
+        if (owner == address(0)) {
+            _safeMint(to, tokenId);
+        } else {
+            require(expires[tokenId] < block.timestamp, "Name already exists and is not expired.");
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:91]: We should use a proper error

File: src/nameservice/ClickResolver.sol [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            revert OffchainLookup(
+                address(this),
+                urls,
+                callData,
+                ClickResolver.resolveWithProof.selector,
+                abi.encode(registry, registrySlot)
+            );
+        } else if (functionSelector == ADDR_MULTICHAIN_SELECTOR) {
+            (, uint256 coinType) = abi.decode(_data[4:], (bytes32, uint256));
+            if (coinType != ZKSYNC_MAINNET_COIN_TYPE) {
```
Comment (ETeissonniere) [inline src/nameservice/ClickResolver.sol:213]: (discussed today, adding here for tracking) Shouldn't we have a coin type for testnet too?

File: hardhat.config.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 
 import { HardhatUserConfig } from "hardhat/config";
 
+import "hardhat-storage-layout";
```
Comment (ETeissonniere) [inline hardhat.config.ts:4]: What is this for?

File: clk-gateway/package.json [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "ts-node": "^10.9.2"
+  },
+  "scripts": {
+    "fmt": "prettier --write src/**/*.ts",
+    "test": "jest",
+    "start": "node dist/index.js",
+    "dev": "nodemon src/index.ts",
+    "build": "tsc",
+    "serve": "yarn build && yarn start",
+    "createToken": "node dist/createToken.js"
```
Comment (ETeissonniere) [inline clk-gateway/package.json:32]: We should use `tsnode` instead of the `node dist/...` pattern

File: clk-gateway/package.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "clk-gateway",
+  "version": "1.0.0",
+  "main": "index.js",
+  "license": "MIT",
```
Comment (ETeissonniere) [inline clk-gateway/package.json:5]: BSD 3 clear clause please :)

**Review by ETeissonniere — CHANGES_REQUESTED** [review]

File: src/nameservice/ClickNameService.sol [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        _grantRole(REGISTERER_ROLE, registrar);
+    }
+
+    /// @inheritdoc IClickNameService
+    function resolve(string memory name) external view returns (address) {
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = ownerOf(tokenId);
+        if (expires[tokenId] <= block.timestamp) {
+            revert NameExpired(owner, expires[tokenId]);
+        }
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:73]: By default, names should be permanent unless we enable expiry via an admin function

File: src/nameservice/ClickNameService.sol [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            revert NameMustBeAlphanumeric();
+        }
+
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = _ownerOf(tokenId);
+        if (owner == address(0)) {
+            _safeMint(to, tokenId);
+        } else {
+            if (expires[tokenId] > block.timestamp) {
+                revert NamAlreadyExists(owner, expires[tokenId]);
```
Comment (ETeissonniere) [inline src/nameservice/ClickNameService.sol:124]:
```suggestion
                revert NameAlreadyExists(owner, expires[tokenId]);
```

File: clk-gateway/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+      if (!result.isEmpty()) {
+        res.status(400).json(result.array());
+        return;
+      }
+      const data = matchedData(req);
+
+      const sub = data.name.split(".")[0];
+      const owner = getAddress(data.owner);
+
+      const response = await clickNameServiceContract.register(owner, sub);
```
Comment (ETeissonniere) [inline clk-gateway/src/index.ts:493]: This is where we might want to do manual nonce management (future ticket?)

File: src/nameservice/ClickNameService.sol (hunk 1 above)
Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: Done

Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: Let's use a big number for expiration of early names.

File: src/nameservice/ClickNameService.sol (hunk 4 above)
Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: done

File: src/nameservice/ClickResolver.sol (hunk 5 above)
Comment (aliXsed) [reply src/nameservice/ClickResolver.sol:142]: This ADDR_MULTICHAIN_SELECTOR path is not used for us now. So no need to do anything for testnet at the moment.

File: hardhat.config.ts (hunk 6 above)
Comment (aliXsed) [reply hardhat.config.ts:4]: The command it enables is `npx hardhat storage-layout`

File: clk-gateway/package.json (hunk 7 above)
Comment (aliXsed) [reply clk-gateway/package.json]: done

File: clk-gateway/package.json (hunk 8 above)
Comment (aliXsed) [reply clk-gateway/package.json]: done

File: src/nameservice/ClickNameService.sol (hunk 3 above)
Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: done

File: src/nameservice/ClickNameService.sol (hunk 2 above)
Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: done

File: clk-gateway/src/index.ts (hunk 11 above)
Comment (aliXsed) [reply clk-gateway/src/index.ts:493]: No manual nonce management just buffering the request and batching them to reduce the likelihood of concurrent registers

File: src/nameservice/ClickNameService.sol (hunk 10 above)
Comment (aliXsed) [reply src/nameservice/ClickNameService.sol]: done

File: clk-gateway/src/index.ts (hunk 11 above)
Comment (aliXsed) [reply clk-gateway/src/index.ts:496]: https://linear.app/nodle/issue/CHA-546

**Pull request merged**